csharp/src/Ice/OutgoingFrame.cs

```diff
@@ -186,176 +156,104 @@ public CompressionResult CompressPayload()
-            {
-                _binaryContextOstr = new OutputStream(Encoding.V20, Data, PayloadEnd);
-                _binaryContextOstr.WriteByteSpan(stackalloc byte[2]); // 2-bytes size place holder
-            }
-            return _binaryContextOstr;
-        }
+            int size = 0;
 
-        // Finish prepares the frame for sending and adjusts the last written segment to match the offset of the written
-        // data. If the frame contains a binary context, Finish appends the entries from defaultBinaryContext (if any)
-        // and rewrites the binary context dictionary size.
-        internal virtual void Finish()
-        {
-            if (!IsSealed)
-            {
-                ArraySegment<byte> defaultBinaryContext = GetDefaultBinaryContext();
+            OutputStream.Position start = ostr.StartFixedLengthSize(sizeLength);
 
-                if (_binaryContextOstr is OutputStream ostr)
-                {
-                    Debug.Assert(_binaryContextKeys != null);
-                    Data[^1] = Data[^1].Slice(0, ostr.Tail.Offset);
-                    if (defaultBinaryContext.Count > 0)
-                    {
-                        // Add segment for each slot that was not written yet.
-                        var istr = new InputStream(defaultBinaryContext, Encoding.V20);
-                        int dictionarySize = istr.ReadSize();
-                        for (int i = 0; i < dictionarySize; ++i)
-                        {
-                            int startPos = istr.Pos;
-                            int key = istr.ReadVarInt();
-                            int entrySize = istr.ReadSize();
-                            istr.Skip(entrySize);
-                            if (!ContainsKey(key))
-                            {
-                                Data.Add(defaultBinaryContext.Slice(startPos, istr.Pos - startPos));
-                                AddKey(key);
-                            }
-                        }
-                    }
+            // First write the overrides, then the InitialBinaryContext entries that were not overridden.
 
-                    ostr.RewriteFixedLengthSize20(_binaryContextKeys.Count, PayloadEnd, 2);
+            if (_binaryContextOverride is Dictionary<int, Action<OutputStream>> binaryContextOverride)
+            {
+                foreach ((int key, Action<OutputStream> action) in binaryContextOverride)
+                {
+                    ostr.WriteVarInt(key);
+                    OutputStream.Position startValue = ostr.StartFixedLengthSize(2);
+                    action(ostr);
+                    ostr.EndFixedLengthSize(startValue, 2);
+                    size++;
                 }
-                else
+            }
+            foreach ((int key, ReadOnlyMemory<byte> value) in InitialBinaryContext)
+            {
+                if (_binaryContextOverride == null || !_binaryContextOverride.ContainsKey(key))
```

Comment:
```suggestion
                if (!(_binaryContextOverride?.ContainsKey(key) ?? false))
```

Reply:
I find the suggested version harder to understand!